fsspec/asyn.py

```diff
@@ -239,20 +239,35 @@ async def _run_coros_in_chunks(
         batch_size = len(coros)
 
     assert batch_size > 0
-    results = []
-    for start in range(0, len(coros), batch_size):
-        chunk = [
-            asyncio.Task(asyncio.wait_for(c, timeout=timeout))
-            for c in coros[start : start + batch_size]
-        ]
-        if callback is not DEFAULT_CALLBACK:
-            [
-                t.add_done_callback(lambda *_, **__: callback.relative_update(1))
-                for t in chunk
-            ]
-        results.extend(
-            await asyncio.gather(*chunk, return_exceptions=return_exceptions),
-        )
+
+    async def _run_coro(coro, i):
+        try:
+            return await asyncio.wait_for(coro, timeout=timeout), i
+        except Exception as e:
+            if not return_exceptions:
+                raise
+            return e, i
+        finally:
+            callback.relative_update(1)
+
+    i = 0
+    n = len(coros)
+    results = [None] * n
+    pending = set()
+
+    while pending or i < n:
+        while len(pending) < batch_size and i < n:
+            pending.add(asyncio.ensure_future(_run_coro(coros[i], i)))
+            i += 1
+
+        if not pending:
+            break
```

Comment:
I wonder whether this condition is actually needed. :thinking:

The outer `while` loop continues only if `pending` is not empty or there are still coroutines left to add to `pending`. If `pending` is empty, then the inner `while` loop will _always_ add at least one coroutine to `pending` because `len(pending) < batch_size` is always true (because `pending` is empty and `batch_size > 0` according to the assertions further above). And if `pending` is not empty, then `i < n` may not hold (i.e. no coroutines are left), but then `if not pending` is `False` anyway.

Reply:
I doubt it hurts - we can check in coverage whether the line ever gets hit; but very large scale calls may behave differently and reveal corner cases. It's worth remembering that function may get called simultaneously from multiple threads.